### DISCUSSION TITLE
Auth implementation ignores built-in role memberships

### DISCUSSION THREAD

**Discussion by howesteve** [supabase/supabase#3107] [Questions] [answered]
Postgresql has a built-in role membership model including a wonderful function pg_has_role() that allows checking for roles membership. More on that here:

https://pgpedia.info/p/pg_has_role.html 

But in supabase, except if I missed something, there is no 1<=>1 mapping between an auth.users user and a built-in pg_user user. That is a design decision that ended up rendering memberships and pg_has_role() useless, and I cannot see any alternatives. All GRANT xxx TO yyy functionality is ignored, and we are expected to implement all security policies/RLS using auth.users.id. No memberships at all. 

So, how can one implement hierarchy checking on RLS policies using supabase? For an example on this built-in functionality using pg_has_role(), check the example below:

https://www.compose.com/articles/row-level-security-with-postgresql-95/

**Comment by steve-chavez** [comment, marked as answer]
@howesteve Having [one role per web user](https://postgrest.org/en/v8.0/auth.html#roles-for-each-web-user) has a significant performance impact on PostgreSQL - this is something that was tried and benchmarked in PostgREST many years ago.

> With 10K roles the PostgREST tests became unbearably slow.

From https://github.com/PostgREST/postgrest/pull/311#issuecomment-147569321

> https://www.compose.com/articles/row-level-security-with-postgresql-95/

At the time of this article(since it mentions postgresql 9.5, it could be from 2016) this was likely not benchmarked.

---

Having said that, [different roles for grouping many web users](https://postgrest.org/en/v8.0/auth.html#web-users-sharing-role) is feasible. Right now Supabase only supports the `authenticated` and `anon` groups but we'll later support adding your own custom groups(this already works in PostgREST but needs additional work for Kong).

**Reply by howesteve** [reply]
Thanks for the reply. 

Actually I read those threads. Let's check what is wrong with them...

1) The [mentioned](PostgREST/postgrest#311 (comment)) tests are not available and are not conclusive; even the author says he should "dig deeper". 
2) That thread was six years ago. RLS was introduced on PostgreSQL 9.5, and its release date was in Jan/2016. But the thread and the so-called "tests" are from Oct/2015! So that was benchmarked on a BETA version, no source code available... and was just assumed that as a fact in the current implementation of supabase.
3) In the [following linked thread](https://github.com/PostgREST/postgrest-docs/issues/170) you have acknowledged "This needs further investigation" (and I agree).

... and what is wrong with current supabase design:
1) Role memberships are completely ignored (show me an RLS select policy for showing all rows that belong to a group)
2) current_user, a standard sql feature, will not be working as expected.
3) The current auth system certainly is slower than it should because currently you must make a lookup to another table in the policy (auth.users->id). Nothing could be faster than just using current_user - no lookups at all! Just as designed and thoroughly documented in the PostgreSQL docs. I have no idea why in implementation would add an additional lookup and kill memberships/pg_has_role()/has_type_privilege/etc. etc. by doing that.
4) As per this [comment ](https://github.com/PostgREST/postgrest/pull/311#issuecomment-147579706) in the same thread, best practices are a 1:1 mapping between db and application.

I bet a benchmark would prove (but I'm not doing it) there is no problem at all with performance using built-in roles - it is designed just for that by the best experts in industry - and actually I have two projects running that way. One of them should have > 40.000 users by now and performance is absolutely fine. In that application, I also have a detailed users table (similar to auth.users) but I do create a system role column matching 1:1 every user created and add that to the detail table. After authentication, the JWT token will have the system role name and I just use SET SESSION AUTHORIZATION xxx; before running a command on the same pooled connection. Everything runs smoothly and intuitively.

How exactly will be this future implementation that will allow "adding your own custom groups"? Will it use plain PostgreSQL roles or another non-standard method that we should be aware?

Honestly, I'm so disappointed. This is a huge design omission from the project, and I found it as soom as I started to review supabase. Now I guess I'll see much more like this I'm going to see...

**Reply by steve-chavez** [reply]
Thanks for the feedback. To put it in actionable terms: we need a new benchmark with at least 40K roles. 

> As per this comment in the same thread, best practices are a 1:1 mapping between db and application.

Note that that comment is mentioning Oracle. If the benchmark indeed proves there's no performance penalty for postgres, then we'd be fine recommending a 1:1 mapping for users that want it.

> How exactly will be this future implementation that will allow "adding your own custom groups"? Will it use plain PostgreSQL roles or another non-standard method that we should be aware?

Plain postgres roles. As mentioned before, this is already working for the PostgREST layer, but for Supabase, currently the JWTs passed are whitelisted in Kong(perhaps this restriction can be lifted).

**Reply by howesteve** [reply]
Thanks for the answer and for the internal implementation details. 
This is not just a matter of taste; it will bring considerable unload to the servers and allow built-in membership/permissions to be used in much more complex ways than are possible now.

Hoping this gets implemented soon since my projects all have an inherited roles system that depends on this.
